token-tracker/src/truth-sources/lp-tokens.js

```diff
@@ -0,0 +1,267 @@
+   * */
+  async findStakers(pairName) {
+    const lpRewardsContractAddress = this.liquidityStaking.lpRewardsContract
+      .options.address
+
+    logger.info(
+      `looking for Transfer events emitted from ${lpRewardsContractAddress} ` +
+        `to ${pairName} pair ${this.liquidityStaking.lpTokenContract.options.address} ` +
+        `between blocks ${this.liquidityStaking.lpCreationBlock} and ${this.targetBlock}`
+    )
+
+    const events = await getPastEvents(
+      this.context.web3,
+      this.liquidityStaking.lpTokenContract,
+      "Transfer",
+      this.liquidityStaking.lpCreationBlock,
+      this.targetBlock
+    )
+    logger.info(`found ${events.length} lp ${pairName} token transfer events`)
+
+    const lpTokenStakersSet = new Set()
+    events.forEach((event) => {
+      // include accounts that staked in LPReward contract only
+      if (event.returnValues.to == lpRewardsContractAddress) {
+        lpTokenStakersSet.add(event.returnValues.from)
+      }
+    })
+
+    logger.info(`found ${lpTokenStakersSet.size} unique historic stakers`)
+
+    return Array.from(lpTokenStakersSet)
+  }
+
+  /**
+   * Retrieves balances of LP KEEP-ETH / KEEP-TBTC pair for stakers in LPRewards* contract
+   *
+   * @param {Array<Address>} lpStakers LP KEEP-ETH / KEEP-TBTC stakers
+   *
+   * @return {Map<Address,BN>} LP Balances by lp stakers
+   */
+  async getLpTokenStakersBalances(lpStakers) {
+    const lpBalanceByStaker = new Map()
+    let expectedTotalSupply = new BN(0)
+
+    for (let i = 0; i < lpStakers.length; i++) {
+      const lpBalance = new BN(
+        await this.liquidityStaking.lpRewardsContract.methods
+          .balanceOf(lpStakers[i])
+          .call({}, this.targetBlock)
+      )
+      if (!lpBalance.isZero()) {
+        lpBalanceByStaker.set(lpStakers[i], lpBalance)
+        expectedTotalSupply = expectedTotalSupply.add(lpBalance)
+      }
+    }
+    const actualTotalSupply = new BN(
+      await this.liquidityStaking.lpRewardsContract.methods
+        .totalSupply()
+        .call({}, this.targetBlock)
+    )
+
+    if (!expectedTotalSupply.eq(actualTotalSupply)) {
+      logger.error(
+        `Sum of LP staker balances ${expectedTotalSupply} does not match the total supply ${actualTotalSupply}`
+      )
+    }
+
+    logger.info(`Total supply of LP Token: ${expectedTotalSupply.toString()}`)
+
+    return lpBalanceByStaker
+  }
+
+  /**
+   * Calculates KEEP for all LP KEEP-ETH / KEEP-TBTC stakers.
+   *
+   * @param {Map<Address, BN>} stakersBalances LP KEEP-ETH / KEEP-TBTC Token amounts by stakers
+   *
+   * @return {Map<Address,BN>} KEEP Tokens in LP KEEP-ETH / KEEP-TBTC at the target block
+   */
+  async calcKeepInStakersBalances(stakersBalances) {
+    logger.info(`check token stakers at block ${this.targetBlock}`)
+
+    const keepInLpByStakers = new Map()
+
+    // Retrieve current pair data
+    const pairData = await getPairData(this.liquidityStaking.lpPairAddress)
```

Comment:
Let's log some of the fetched data that we will use later.

Reply:
Do you think we should log under `debug` or `info` level? The thing is that in the loop which is under we log `logger.info("Staker: ${stakerAddress} - LP Balance: ${lpBalance} - KEEP in LP: ${keepInLPToken}")` and this provides good information on staker-balances. I just don't want to cross the point when we'll have too much logs.

Reply:
This will give us just one additional line, we're not looping this call, right? I'm fine with either one of info or debug.

Reply:
This is actually gone now 09609e6, I'll log new data